File: bst.go
package trees

type Node struct {
	item  int
	left  *Node
	right *Node
}

func (n Node) NewNode(item int) *Node {
	return &Node{
		item:  item,
		left:  nil,
		right: nil,
	}
}

func Insert(root *Node, key int) {
	if root == nil {
		root.NewNode(key)
	}
	if root.item < key {
		Insert(root.right, key)
	} else {
		Insert(root.left, key)
	}
}

func InorderElement(cur *Node) *Node {
	for cur.left != nil {
		cur = cur.left
	}
	return cur
}

func Delete(root *Node, key int) *Node {
	if root == nil {
		return nil
	} else if root.item > key {
		Delete(root.left, key)
	} else if root.item < key {
		Delete(root.right, key)
	} else {
		if root.left == nil {
			root = root.right
			root.right = nil
		} else if root.right == nil {
			root = root.left
			root.left = nil
		} else {
			d := InorderElement(root.right)
			root.item = d.item
			d = nil
		}
	}
	return root
}

func Search(root *Node, key int) bool {
	if root == nil {
		return false
	} else if root.item > key {
		Search(root.left, key)
	} else if root.item < key {
		Search(root.right, key)
	} else {
		return true
	}
	return false
}

File: bst.ts
interface Node {
    data: number,
    left: Node | null,
    right: Node | null
}

class BST {

    constructor() {
    }

    newNode(val: number): Node {
        return {
            data: val,
            left: null,
            right: null
        } as Node;
    }

    insertNode(val: number, root: Node | null): Node {
        if (root == null) {
            return this.newNode(val);        
        }
        if (val > root?.data) {
            root.right = this.insertNode(val, root.right);
        } else {
            root.left = this.insertNode(val, root.left);
        } 
        return root;
    }

    inorder(root: Node | null) {
        if (root == null) {
            return;
        }
        this.inorder(root?.left);
        console.log(root.data);
        this.inorder(root.right);
    }

    checkBST(root: Node | null): boolean {
        if (root?.left != null) {
            if (root.data < root.left.data) {
                return false;
            }
        }
        return true;
    }

    height(root: Node | null, h: number): number {
        if (root == null) {
            return h;
        }
        return Math.max(this.height(root.left, h+1), this.height(root.right, h+1));
    }
}

let node: Node;
const bst: BST = new BST();
node = bst.insertNode(23, null);
node = bst.insertNode(43, node);
node = bst.insertNode(2, node);
node = bst.insertNode(98, node);
node = bst.insertNode(67, node);
node = bst.insertNode(4, node);
//bst.inorder(node);
console.log(bst.height(node, 0));






